### PR TITLE
[AIRFLOW-3623] Support download logs by attempts from UI

### PR DESCRIPTION
**Pull request by pingzh**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3623
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Support download logs by attempts from UI

image
image
image
image

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

@KevinYang21

### CODE REVIEW

**Comment by feng-tao** [comment]
Does the feature work when the log is in a remote location(like some s3 directory)?

**Comment by pingzh** [comment]
> Does the feature work when the log is in a remote location(like some s3 directory)?

Hi @feng-tao , that is a great question. It should work since my changes do not involve any logger handlers.

**Comment by pingzh** [comment]
> Does the feature work when the log is in a remote location(like some s3 directory)?

Hi @feng-tao I have tested it against the elasticsearch log handler. It works fine.

File: airflow/utils/helpers.py [hunk 1]
```diff
@@ -297,3 +297,16 @@ def parse_template_string(template_string):
         return None, Template(template_string)
     else:
         return template_string, None
+
+
+def render_log_filename(ti, try_number, filename_template):
```
Comment (feng-tao) [inline airflow/utils/helpers.py:302]: nit: please provides a docstring for the function.

File: airflow/www/templates/airflow/dag.html [hunk 2]
```diff
@@ -320,6 +326,29 @@ <h4 class="modal-title" id="dagModalLabel">
             $("#div_btn_subdag").show();
             subdag_id = "{{ dag.dag_id }}."+t;
         }
+
+      $("#try_index > li").remove()
```
Comment (feng-tao) [inline airflow/www/templates/airflow/dag.html]: nit: semicolon in the end?

File: airflow/www/templates/airflow/dag.html [hunk 3]
```diff
@@ -320,6 +326,29 @@ <h4 class="modal-title" id="dagModalLabel">
...
+      var startIndex = (try_numbers > 2 ? 0 : 1)
```
Comment (feng-tao) [inline airflow/www/templates/airflow/dag.html]: nit: semicolon in the end?

File: airflow/www/templates/airflow/dag.html [hunk 4]
```diff
@@ -320,6 +326,29 @@ <h4 class="modal-title" id="dagModalLabel">
...
+      for (var index = startIndex; index <  try_numbers; index++) {
+        var url = "{{ url_for('airflow.get_logs_with_metadata') }}" +
+          "?dag_id=" + encodeURIComponent(dag_id) +
+          "&task_id=" + encodeURIComponent(task_id) +
+          "&execution_date=" + encodeURIComponent(execution_date) +
+          "&metadata=null" +
+          "&format=file"
```
Comment (feng-tao) [inline airflow/www/templates/airflow/dag.html]: nit: semicolon in the end?

File: airflow/www/views.py [hunk 5]
```diff
@@ -770,7 +774,12 @@ def get_logs_with_metadata(self, session=None):
         task_id = request.args.get('task_id')
         execution_date = request.args.get('execution_date')
         dttm = pendulum.parse(execution_date)
-        try_number = int(request.args.get('try_number'))
+        if request.args.get('try_number', None) is not None:
```
Comment (feng-tao) [inline airflow/www/views.py]: small nit: request.args.get('try_number', None) could be request.args.get('try_number') as the default should be None

File: airflow/www_rbac/templates/airflow/dag.html [hunk 6]
```diff
@@ -314,12 +320,35 @@ <h4 class="modal-title" id="dagModalLabel">
...
-            $("#div_btn_subdag").show();
-            subdag_id = "{{ dag.dag_id }}."+t;
+      if (subdag_id===undefined)
+          $("#div_btn_subdag").hide();
+      else {
+          $("#div_btn_subdag").show();
+          subdag_id = "{{ dag.dag_id }}."+t;
+      }
+
+      $("#try_index > li").remove()
```
Comment (feng-tao) [inline airflow/www_rbac/templates/airflow/dag.html]: same for the semicolon

**Review by feng-tao — COMMENTED** [review]
thanks. Just left some small comments.

File: airflow/utils/helpers.py (hunk 1 above)
Comment (pingzh) [reply airflow/utils/helpers.py:302]: 👍

File: airflow/www/templates/airflow/dag.html (hunk 2 above)
Comment (pingzh) [reply airflow/www/templates/airflow/dag.html]: 👍

File: airflow/www/templates/airflow/dag.html (hunk 3 above)
Comment (pingzh) [reply airflow/www/templates/airflow/dag.html]: 👍

File: airflow/www/templates/airflow/dag.html (hunk 4 above)
Comment (pingzh) [reply airflow/www/templates/airflow/dag.html]: 👍

File: airflow/www_rbac/templates/airflow/dag.html (hunk 6 above)
Comment (pingzh) [reply airflow/www_rbac/templates/airflow/dag.html]: :+1

File: airflow/www/views.py (hunk 5 above)
Comment (pingzh) [reply airflow/www/views.py]: 👍

**Comment by pingzh** [comment]
Thanks for your feedback, @feng-tao.  I have addressed them.

**Comment by feng-tao** [comment]
thanks @pingzh

**Comment by msumit** [comment]
@pingzh @feng-tao sorry for late comment, but won't it be better if we provide download logs feature at the "View Logs" page itself, instead of from the model dialog? As IMO people would anyway want to view logs first and then take the decision of downloading it?

**Comment by pingzh** [comment]
@msumit great point. One big reason for this feature is for large log files, it is expensive to pull the large logs and render in the web UI.

**Comment by msumit** [comment]
@pingzh totally agree with the problem of large logs, but I think that's a diff problem to solve, and can be solved easily, for example, see this https://github.com/apache/incubator-airflow/pull/3992.

**Comment by feng-tao** [comment]
@msumit , I think your point is valid, but sometimes the log file is very hard to render and view if the file is too large(not sure of the status or the implementation detail for #3992). The log file is very critical for debugging production issue on why the task not running. And adding this action provides to task instance UI panel(all the operations for task instance) provides a way to download the file based on different try number. Hence I think this pr provides value on solving this issue.

**Comment by pingzh** [comment]
@msumit Thanks for the info. In my mind, for large log files, `tailing` or `range` lines does not provide lots of insights when trying to find something. And it looks like #3992 only supports file log handler at least for now.

It is also a good idea to add a download btn after users are in the "View Logs" page.

**Comment by kaxil** [comment]
Can you please resolve an error that was in the PR but because of broken Travis it wasn't detected:

```
======================================================================
46) ERROR: test_get_logs_with_metadata_as_download_file (tests.www_rbac.test_views.TestLogView)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/www_rbac/test_views.py line 569 in test_get_logs_with_metadata_as_download_file
      response = self.app.get(url)
   AttributeError: 'Flask' object has no attribute 'get'
```
Please fix it in the new PR.

**Comment by pingzh** [comment]
@kaxil sorry about that. Here is the PR to fix it: https://github.com/apache/airflow/pull/4446

> Can you please resolve an error that was in the PR but because of broken Travis it wasn't detected:
> 
> ```
> ======================================================================
> 46) ERROR: test_get_logs_with_metadata_as_download_file (tests.www_rbac.test_views.TestLogView)
> ----------------------------------------------------------------------
>    Traceback (most recent call last):
>     tests/www_rbac/test_views.py line 569 in test_get_logs_with_metadata_as_download_file
>       response = self.app.get(url)
>    AttributeError: 'Flask' object has no attribute 'get'
> ```
> Please fix it in the new PR.

**Pull request merged**